push/service.go

```diff
@@ -39,68 +39,165 @@ type Headers struct {
+	ErrServiceUnavailable  = errors.New("the service is unavailable")
+
+	// HTTP Status errors.
+
+	ErrBadRequest = errors.New("bad request")
+	ErrGone       = errors.New("the device token is no longer active for the topic")
+	ErrUnknown    = errors.New("unknown error")
 )
 
+var errorReason = map[string]error{
+	"PayloadEmpty":              ErrPayloadEmpty,
+	"PayloadTooLarge":           ErrPayloadTooLarge,
+	"BadTopic":                  ErrBadTopic,
+	"TopicDisallowed":           ErrTopicDisallowed,
+	"BadMessageId":              ErrBadMessageID,
+	"BadExpirationDate":         ErrBadExpirationDate,
+	"BadPriority":               ErrBadPriority,
+	"MissingDeviceToken":        ErrMissingDeviceToken,
+	"BadDeviceToken":            ErrBadDeviceToken,
+	"DeviceTokenNotForTopic":    ErrDeviceTokenNotForTopic,
+	"Unregistered":              ErrUnregistered,
+	"DuplicateHeaders":          ErrDuplicateHeaders,
+	"BadCertificateEnvironment": ErrBadCertificateEnvironment,
+	"BadCertificate":            ErrBadCertificate,
+	"Forbidden":                 ErrForbidden,
+	"BadPath":                   ErrBadPath,
+	"MethodNotAllowed":          ErrMethodNotAllowed,
+	"TooManyRequests":           ErrTooManyRequests,
+	"IdleTimeout":               ErrIdleTimeout,
+	"Shutdown":                  ErrShutdown,
+	"InternalServerError":       ErrInternalServerError,
+	"ServiceUnavailable":        ErrServiceUnavailable,
+	"MissingTopic":              ErrMissingTopic,
+}
+
 type response struct {
+	// Reason for failure
 	Reason string `json:"reason"`
-	// timestamp, other fields?
+	// Timestamp for 410 errors (maybe this is an int)
+	Timestamp string `json:"timestamp"`
 }
 
+const statusTooManyRequests = 429
+
 // Push notification to APN service after performing serialization.
-func (s *Service) Push(deviceToken string, headers *Headers, payload json.Marshaler) error {
+func (s *Service) Push(deviceToken string, headers *Headers, payload interface{}) (string, error) {
 	b, err := json.Marshal(payload)
 	if err != nil {
-		return err
+		return "", err
 	}
 	return s.PushBytes(deviceToken, headers, b)
 }
 
 // PushBytes notification to APN service.
-func (s *Service) PushBytes(deviceToken string, headers *Headers, payload []byte) error {
+func (s *Service) PushBytes(deviceToken string, headers *Headers, payload []byte) (string, error) {
 	urlStr := fmt.Sprintf("%v/3/device/%v", s.Host, deviceToken)
 
 	req, err := http.NewRequest("POST", urlStr, bytes.NewReader(payload))
 	if err != nil {
-		return err
+		return "", err
 	}
 	req.Header.Set("Content-Type", "application/json")
 	headers.set(req)
 
 	resp, err := s.Client.Do(req)
 	if err != nil {
-		return err
+		return "", err
 	}
 	defer resp.Body.Close()
 
 	if resp.StatusCode == http.StatusOK {
-		return nil
+		return resp.Header.Get("apns-id"), nil
 	}
 
 	// read entire response body
 	// TODO: could decode while reading instead
 	body, err := ioutil.ReadAll(resp.Body)
```

Comment:
#picky from this line down I don't see any test coverage. Would be good to test at least one of the error cases.